[audio][android][ios][web] Fix recordForDuration status updates and add new record() options API

hirbod opened the pull request:
# Why

<!--
Please describe the motivation for this PR, and link to relevant GitHub issues, forums posts, or feature requests.
-->

  - recordForDuration() and startRecordingAtTime() methods had inconsistent behavior across platforms
  - No unified API for recording options - required separate method calls
  - iOS atTime parameter was incorrectly treated as relative delay instead of absolute device time
  - Android and Web didn't emit completion events, causing demo UI to not show "last recording"
  - Android required manual re-preparation between recordings, unlike iOS/Web
  - Manual recording stops on Android didn't emit events on correct thread context

Fixes #38565

# How

<!--
How did you build this feature or fix this bug and why?
-->

  New API:
  - Added record(options?: RecordingStartOptions) with atTime and forDuration properties
  - Deprecated recordForDuration() and startRecordingAtTime() methods (maintained for backward compatibility)
  - Added TypeScript types RecordingStartOptions with platform-specific documentation

  iOS:
  - Fixed atTime to use absolute device time (deviceCurrentTime + atTime) instead of relative delay
  - Added state synchronization for native AVAudioRecorder calls
  - Full support for both atTime and forDuration options

  Android:
  - Added completion event emission in stopRecording() with proper thread context (mainQueue)
  - Implemented auto-preparation after recording stops to match iOS/Web behavior
  - Made atTime a no-op (iOS-only feature) with clear documentation and comments
  - Maintained forDuration support with coroutine timers

  Web:
  - Made atTime a no-op (iOS-only feature) with clean conditional removal
  - Maintained forDuration support with setTimeout

  Demo:
  - Added toggle switches for testing new record({ atTime: 3, forDuration: 5 }) API
  - Added automatic completion detection for forDuration recordings
  - Updated UI to show platform-specific behavior ("iOS only" labels)

# Test Plan

<!--
Please describe how you tested this change and how a reviewer could reproduce your test, especially if this PR does not include automated tests! If possible, please also provide terminal output and/or screenshots demonstrating your test/reproduction.
-->

bare-expo on Android, iOS and Web

# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
-->

- [x] I added a `changelog.md` entry and rebuilt the package sources according to [this short guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

hirbod committed:
wip

fix linting

fix linting

fix state

fix demo

correctly handle atTime recording using deviceCurrentTime

fix

Enhance AudioRecorder functionality by capturing file path on stop and preparing for next recording. Emit completion event with updated status including URL. Clean up file path after reset.

chore: missing changelog

fix: linting

fix: comment/spotless apply

improve atTime documentation

alanjhughes commented on packages/expo-audio/android/src/main/java/expo/modules/audio/AudioModule.kt:
```diff
@@ -465,10 +465,12 @@ class AudioModule : Module() {
       }
 
-      Function("record") { recorder: AudioRecorder ->
+      Function("record") { recorder: AudioRecorder, options: Map<String, Any>? ->
```
```suggestion
      Function("record") { recorder: AudioRecorder, options: RecordOptions? ->
```

```swift
struct RecordOptions: Record {
  @Field var forDuration: Double?
  @Field var atTime: Double?
}
```
Let's create a record for this

alanjhughes commented on packages/expo-audio/ios/AudioModule.swift:
```diff
@@ -234,8 +234,30 @@ public class AudioModule: Module {
       }
 
-      Function("record") { (recorder: AudioRecorder) -> [String: Any] in
+      Function("record") { (recorder: AudioRecorder, options: [String: Any]?) in
```
```suggestion
      Function("record") { (recorder: AudioRecorder, options: RecordOptions?) in
```

alanjhughes commented on packages/expo-audio/ios/AudioModule.swift:
```diff
@@ -234,8 +234,30 @@ public class AudioModule: Module {
-      Function("record") { (recorder: AudioRecorder) -> [String: Any] in
+      Function("record") { (recorder: AudioRecorder, options: [String: Any]?) in
         try checkPermissions()
+        let atTime = options?["atTime"] as? Double
+        let forDuration = options?["forDuration"] as? Double
+
+        if let atTime = atTime, let forDuration = forDuration {
+          // Convert relative delay to absolute device time
+          let absoluteTime = recorder.ref.deviceCurrentTime + TimeInterval(atTime)
+          recorder.ref.record(atTime: absoluteTime, forDuration: TimeInterval(forDuration))
+          recorder.updateStateForDirectRecording()
+          return recorder.getRecordingStatus()
+        }
+        if let atTime = atTime {
+          // Convert relative delay to absolute device time
+          let absoluteTime = recorder.ref.deviceCurrentTime + TimeInterval(atTime)
+          recorder.ref.record(atTime: absoluteTime)
+          recorder.updateStateForDirectRecording()
+          return recorder.getRecordingStatus()
+        }
+        if let forDuration = forDuration {
+          recorder.ref.record(forDuration: TimeInterval(forDuration))
+          recorder.updateStateForDirectRecording()
+          return recorder.getRecordingStatus()
+        }
```
```suggestion
       switch (options?.atTime, options?.forDuration) {
        case let (atTime?, forDuration?):
           // Convert relative delay to absolute device time
          let absoluteTime = recorder.ref.deviceCurrentTime + TimeInterval(atTime)
          recorder.ref.record(atTime: absoluteTime, forDuration: TimeInterval(forDuration))
          recorder.updateStateForDirectRecording()
          return recorder.getRecordingStatus()
        case let (atTime?, nil):
           let absoluteTime = recorder.ref.deviceCurrentTime + TimeInterval(atTime)
          recorder.ref.record(atTime: absoluteTime)
          recorder.updateStateForDirectRecording()
          return recorder.getRecordingStatus()
        case let (nil, forDuration?):
           recorder.ref.record(forDuration: TimeInterval(forDuration))
          recorder.updateStateForDirectRecording()
          return recorder.getRecordingStatus()
        default:
            recorder.startRecording()
        }
```

alanjhughes commented on packages/expo-audio/android/src/main/java/expo/modules/audio/AudioModule.kt:
```diff
@@ -465,10 +465,12 @@ class AudioModule : Module() {
         checkRecordingPermission()
         if (recorder.isPrepared) {
-          recorder.record()
+          val atTime = options?.get("atTime") as? Double
+          val forDuration = options?.get("forDuration") as? Double
+          recorder.recordWithOptions(atTime, forDuration)
```
```suggestion
          recorder.recordWithOptions(options.atTime, options.forDuration)
```

alanjhughes commented on packages/expo-audio/android/src/main/java/expo/modules/audio/AudioRecorder.kt:
```diff
@@ -81,21 +81,34 @@ class AudioRecorder(
+        }
       }
+    } else {
+      record()
```
```suggestion
    forDurationSeconds?.let {
      record()
      recordingTimerJob = appContext?.mainQueue?.launch {
        delay((it * 1000).toLong())
        if (isRecording || isPaused) {
          stopRecording()
        }
      }
    } ?: record()
```

alanjhughes commented on packages/expo-audio/ios/AudioRecorder.swift:
```diff
@@ -112,12 +112,22 @@ class AudioRecorder: SharedRef<AVAudioRecorder>, RecordingResultHandler {
     }
 
-    ref.record()
+    _ = ref.record()
```
Is there a warning without this?

hirbod replied:
false alarm

alanjhughes reviewed (changes requested):
Looks good! Just left some comments

hirbod committed:
address pr nits

fix broken expo-go podfile

Revert "fix broken expo-go podfile"

This reverts commit 32a86788cb9e923ee44af68e4422fb19c95e2e2d.

Update packages/expo-audio/src/Audio.types.ts

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

docs pr nits

amandeepmittal commented on packages/expo-audio/src/Audio.types.ts:
```diff
@@ -275,6 +275,36 @@ export enum AudioQuality {
+   * The time in seconds to wait before starting the recording.
+   * If not provided, recording starts immediately.
+   *
+   * **Platform behavior:**
+   * - iOS: Uses native AVAudioRecorder.record(atTime:) for precise timing.
+   *   ⚠️ **Important**: The recording process starts immediately (you'll see status updates),
+   *   but actual audio capture begins after the specified delay. This is not a countdown -
+   *   the recorder is active but silent during the delay period.
+   * - Android: Ignored, recording starts immediately
+   * - Web: Ignored, recording starts immediately
```
```suggestion
   * **Platform behavior:**
   * - Android: Ignored, recording starts immediately
   * - iOS: Uses native AVAudioRecorder.record(atTime:) for precise timing.
   * - Web: Ignored, recording starts immediately
   *
   * > **warning** On iOS, the recording process starts immediately (you'll see status updates),
   * but actual audio capture begins after the specified delay. This is not a countdown, since
   * the recorder is active but silent during the delay period.
```

Let's use the platform order as per the writing style guide: https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md#platform-order-when-referencing-android-ios-and-web-or-multiple-platforms

I'd avoid the emoji pattern you are using to call out the warning. Let's use the standard callout pattern after the platform behavior section and use `> **warning**` instead of important (we use important for experimental feature indication mostly). If this warning is only related to iOS, in the callout, let's mention the platform name.

amandeepmittal replied:
Also, since this marked as platform iOS only, I think Android and Web info might not be required.
